File: unrated/typical90_q.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using matll = vector<vector<ll>>;

#define sort_all(v) sort(v.begin(), v.end())
#define rep(i, n) for (ll i = 0; i < ll(n); i++)
#define for_itr(id, itr) for (auto& id : itr)

class Fenwicktree {
   private:
    vector<ll> tree;

   public:
    ll N;
    Fenwicktree(int n0) {
        N = n0;
        tree = vector<ll>(N, 0);
    }

    ll accsum(int r) {
        ll res = 0;
        r++;
        while (r > 0) {
            res += tree[r - 1];
            r -= r & -r;
        }
        return res;
    }

    void update(ll idx, ll x) {
        idx++;
        while (idx <= N) {
            tree[idx - 1] += x;
            idx += idx & -idx;
        }
        return;
    }

    ll query(ll l, ll r) { return accsum(r - 1) - accsum(l - 1); }
};

int main() {
    ll n, m;
    cin >> n >> m;
    matll lr(m, vector<ll>(2, -1));
    rep(i, m) {
        cin >> lr[i][0] >> lr[i][1];
        lr[i][0]--, lr[i][1]--;
        lr[i][1] *= -1;
    }
    sort_all(lr);

    Fenwicktree bit(n);
    ll ans = 0;
    for_itr(lri, lr) {
        ll l = lri[0], r = -lri[1];
        bit.update(r, 1);
        ans += bit.query(l + 1, r);
    }
    cout << ans << endl;
    return 0;
}

File: unrated/typical90_q.py
class Fenwicktree():

    def __init__(self, n):
        self.n = n
        self.tree = [0] * n

    def accsum(self, i):
        i, res = i + 1, 0
        while i > 0:
            res += self.tree[i - 1]
            i -= i & -i
        return res

    def update(self, i, x):
        i += 1
        while i <= self.n:
            self.tree[i - 1] += x
            i += i & -i

    def query(self, i, j):
        return self.accsum(j - 1) - self.accsum(i - 1)


def main():
    n, m = map(int, input().split())
    lr = [list(map(int, input().split())) for _ in range(m)]
    lr = sorted((l-1, -r+1) for l, r in lr)

    bit = Fenwicktree(n)
    ans = 0
    for l, r in lr:
        bit.update(-r, 1)
        ans += bit.query(l+1, -r)
    print(ans)


if __name__ == '__main__':
    main()
